Implement Eclipse Jetty core HTTP handler adapter

gregw opened the pull request:
This provides an implementation of a HTTP Handler Adaptor that is coded directly to the Eclipse Jetty core API, bypassing any servlet implementation.

Fixes #32035

gregw committed:
Initial boiler plate implementation of JettyCore HttpHandler Adaptor

gregw committed:
Initial boiler plate implementation of JettyCore HttpHandler Adaptor

gregw committed:
More implementation of the request side

gregw committed:
More implementation of the request side

gregw committed:
Created JettyCoreHttpServer

gregw committed:
Created JettyCoreHttpServer

gregw committed:
Created JettyCoreHttpServer

gregw committed:
Created JettyCoreHttpServer

gregw committed:
WIP on tests

gregw committed:
WIP on tests

gregw committed:
WIP on tests

gregw committed:
WIP on tests

gregw committed:
Cleanup start/stop for jetty

gregw committed:
Cleanup start/stop for jetty

gregw committed:
retainable bytebuffer!

gregw committed:
retainable bytebuffer!

gregw committed:
Non blocking invocation type

gregw committed:
Non blocking invocation type

gregw committed:
split out into separate files

gregw committed:
split out into separate files

gregw committed:
minor httpHeader optimizations

gregw committed:
minor httpHeader optimizations

gregw committed:
since updated with expected release

gregw committed:
since updated with expected release

lachlan-roberts committed:
Initial implementation of the writeWith methods in JettyCoreServerHttpResponse

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Initial implementation of the writeWith methods in JettyCoreServerHttpResponse

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Implement the ZeroCopyHttpOutputMessage interface for JettyCoreServerHttpResponse

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Implement the ZeroCopyHttpOutputMessage interface for JettyCoreServerHttpResponse

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Add temporary fix for cookies

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Add temporary fix for cookies

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
disable setting of the same site cookie attribute to fix test

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
disable setting of the same site cookie attribute to fix test

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
rearrange methods in JettyCoreServerHttpResponse

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
rearrange methods in JettyCoreServerHttpResponse

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
fix issues with cookies, committing response, and with ZeroCopyHttpOutputMessage

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
fix issues with cookies, committing response, and with ZeroCopyHttpOutputMessage

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
cleanups

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
cleanups

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
cleanups

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
cleanups

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

gregw committed:
removed TODO

gregw committed:
removed TODO

gregw committed:
Reformatted code

gregw committed:
Reformatted code

pivotal-cla commented:
@gregw Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-framework&pullRequestId=32097)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=spring-projects/spring-framework&pullRequestId=32097) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

gregw committed:
Fixed unset response status bug

Jetty has an unset response status of 0, this should be converted to 200 for spring.

gregw committed:
Fixed unset response status bug

Jetty has an unset response status of 0, this should be converted to 200 for spring.

gregw committed:
Request header mutation support

When mutating the request, a mutable HttpFields instance is created.

gregw committed:
Request header mutation support

When mutating the request, a mutable HttpFields instance is created.

gregw committed:
Do not use canonical path

Spring appears to want the unadulterated raw path complete with parameters etc.

gregw committed:
Do not use canonical path

Spring appears to want the unadulterated raw path complete with parameters etc.

gregw commented:
@pivotal-cla Working on getting a CCLA signed. Stand by....

gregw committed:
Track leaks (for now) and retain DataBuffer chunks

gregw committed:
Track leaks (for now) and retain DataBuffer chunks

lachlan-roberts committed:
Fixes for Jetty Core WebSocket

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Fixes for Jetty Core WebSocket

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Temporary fix for cookies in JettyCoreServerHttpResponse with TODOs

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Temporary fix for cookies in JettyCoreServerHttpResponse with TODOs

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Do not actually commit response in JettyCoreServerHttpResponse & fix cookies

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Do not actually commit response in JettyCoreServerHttpResponse & fix cookies

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Add JettyCore upgrade option in HandshakeWebSocketService

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Add JettyCore upgrade option in HandshakeWebSocketService

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
move getNativeRequest/Response to ServerHttpRequest/Response interfaces

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
move getNativeRequest/Response to ServerHttpRequest/Response interfaces

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

gregw committed:
Reformatted code

gregw committed:
Reformatted code

gregw committed:
WIP on fixing leaks

gregw committed:
WIP on fixing leaks

gregw committed:
WIP on fixing leaks

gregw committed:
WIP on fixing leaks

gregw commented on spring-webflux/src/main/java/org/springframework/web/reactive/function/server/DefaultServerRequestBuilder.java:
```diff
@@ -297,6 +297,11 @@ public MultiValueMap<String, String> getQueryParams() {
+
+		@Override
+		public <T> T getNativeRequest() {
+			return null;
```
@lachlan-roberts this could be the ServerRequest passed into the builder?

gregw committed:
WIP on fixing tests

gregw committed:
WIP on fixing tests

gregw committed:
Updated to jetty 12.0.6

gregw committed:
Updated to jetty 12.0.6

gregw committed:
Updated to jetty 12.0.6

gregw commented:
@lachlan-roberts can you sign the individual CLA.

gregw commented:
This PR is failing 2 tests that undertow also fails: See https://github.com/spring-projects/spring-framework/issues/25310.
Thus I'm currently suspecting bad tests or a spring bug, as both Jetty core and undertow are similar fully asynchronous integrations without servlets. We will investigate more, but any ideas that can help...

pivotal-cla commented:
@gregw Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-framework&pullRequestId=32097)!

gregw commented on spring-web/src/main/java/org/springframework/http/server/reactive/JettyCoreServerHttpResponse.java:
```diff
@@ -0,0 +1,397 @@
+		Callback.Completable callback = new Callback.Completable();
+		mono = Mono.fromFuture(callback);
+		try {
+			// TODO: Why does intellij warn about possible blocking call?
+			//       Because it can block and we want to be fully asynchronous.  Use AsynchronousFileChannel
+			SeekableByteChannel channel = Files.newByteChannel(file, StandardOpenOption.READ);
```
`Files.newByteChannel` is a blocking operation, but I see that undertow also use it and just live with the warning.   The chance of it actually blocking for any significant time is vanishingly small and there is no chance of dead lock.

gregw committed:
Move all response cookies to multiMap for possible mutation by spring

gregw committed:
Move all response cookies to multiMap for possible mutation by spring

gregw commented on spring-web/src/main/java/org/springframework/http/server/reactive/JettyRetainedDataBuffer.java:
```diff
@@ -0,0 +1,286 @@
+
+	public JettyRetainedDataBuffer(DataBufferFactory dataBufferFactory, Content.Chunk chunk) {
+		this.chunk = chunk;
+		// this.dataBuffer = dataBufferFactory.wrap(BufferUtil.copy(chunk.getByteBuffer())); // TODO this copy avoids multipart bugs
+		this.dataBuffer = dataBufferFactory.wrap(chunk.getByteBuffer()); // TODO avoid double slice?
```
This is the work around for the failing multipart tests (see #25310)

gregw commented on spring-web/src/main/java/org/springframework/http/server/reactive/JettyCoreServerHttpRequest.java:
```diff
@@ -0,0 +1,219 @@
+		// We access the request body as a Flow.Publisher, which is wrapped as an org.reactivestreams.Publisher and
+		// then wrapped as a Flux.   The chunks are converted to RetainedDataBuffers with wrapping and can be
+		// retained within a call to onNext.
+		return Flux.from(FlowAdapters.toPublisher(Content.Source.asPublisher(this.request))).map(this::wrap).doOnNext(this::release);
```
The doOnNext is wrong.  We are releasing before we pass down stream

gregw replied:
I've replaced with an ugly AtomicReference.

gregw committed:
Use an atomicReference to release the last used databuffer after onNext has returned.

gregw committed:
Use an atomicReference to release the last used databuffer after onNext has returned.

gregw commented on spring-web/src/main/java/org/springframework/http/server/reactive/JettyCoreServerHttpRequest.java:
```diff
@@ -0,0 +1,226 @@
+		// then wrapped as a Flux.   The chunks are converted to RetainedDataBuffers with wrapping and can be
+		// retained within a call to onNext.
+
+		// TODO find a better way to release after each onNext call to a subscriber
+		AtomicReference<DataBuffer> last = new AtomicReference<>();
+		return Flux.from(FlowAdapters.toPublisher(Content.Source.asPublisher(this.request)))
+				.map(this::wrap)
+				.doOnNext(db -> release(last.getAndSet(db)))
+				.doOnComplete(() -> release(last.getAndSet(null)));
```
@lachlan-roberts @sbordet Any ideas on how to make this release a bit less ugly?  I want to release after onNext returns, but the only hooks I have is the next call to onNext or the final call to onComplete

gregw replied:
@lachlan-roberts I think you are correct.  It is the implementation of `onNext(Databuffer db)` that is responsible to do the release.   I bit of a strange contract, but I can see that is how they are doing it elsewhere in the code.   Good one! Thanks!

gregw committed:
less allocations for releasing after onNext

gregw committed:
less allocations for releasing after onNext

gregw commented on spring-web/src/main/java/org/springframework/http/server/reactive/JettyCoreServerHttpRequest.java:
```diff
@@ -0,0 +1,234 @@
+		// retained within a call to onNext.
+
+		// TODO find a better way to release after each onNext call to a subscriber
+		DataBufferReleaser releaser = new DataBufferReleaser();
+		return Flux.from(FlowAdapters.toPublisher(Content.Source.asPublisher(this.request)))
+				.map(this::wrap)
+				.doOnNext(releaser::onNext)
+				.doOnComplete(releaser::onComplete);
```
@lachlan-roberts @sbordet  previous comment still applies to this version.  This is a little better, but holds onto a databuffer until the next one arrives, which by definition prevents the previous buffer being reused.  We really want to release after onNext has been called.

gregw replied:
@lachlan-roberts fixed it!

lachlan-roberts committed:
release DataBuffer from the JettyCoreServerHttpResponse

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
release DataBuffer from the JettyCoreServerHttpResponse

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

gregw committed:
fixed style

gregw committed:
fixed style

olamy committed:
fix build with jetty 12.0.7-SNAPSHOT

Signed-off-by: Olivier Lamy <olamy@apache.org>

olamy committed:
fix build with jetty 12.0.7-SNAPSHOT

Signed-off-by: Olivier Lamy <olamy@apache.org>

lachlan-roberts committed:
Implement a JettyWebSocketSession based on demand.

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Implement a JettyWebSocketSession based on demand.

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Add a Jetty implementation of the spring WebSocketClient interface.

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Add a Jetty implementation of the spring WebSocketClient interface.

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Prevent possible ReadPendingException from multiple demand.

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Prevent possible ReadPendingException from multiple demand.

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Update the JettyWebSocketHandlerAdapter as a Session.Listener

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Update the JettyWebSocketHandlerAdapter as a Session.Listener

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
fixes for checkstyle

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
fixes for checkstyle

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Ensure DataBuffer is processed correctly for BINARY messages.

Signed-off-by: Lachlan Roberts <lachlan.p.roberts@gmail.com>

lachlan-roberts committed:
Ensure DataBuffer is processed correctly for BINARY messages.

Signed-off-by: Lachlan Roberts <lachlan.p.roberts@gmail.com>

lachlan-roberts committed:
Wait for close of Session before calling handlerCompletionSink

Signed-off-by: Lachlan Roberts <lachlan.p.roberts@gmail.com>

lachlan-roberts committed:
Wait for close of Session before calling handlerCompletionSink

Signed-off-by: Lachlan Roberts <lachlan.p.roberts@gmail.com>

lachlan-roberts committed:
fix checkstyle error

Signed-off-by: Lachlan Roberts <lachlan.p.roberts@gmail.com>

lachlan-roberts committed:
fix checkstyle error

Signed-off-by: Lachlan Roberts <lachlan.p.roberts@gmail.com>

gregw commented on framework-api/framework-api.gradle:
```diff
@@ -8,6 +8,7 @@ description = "Spring Framework API Docs"
 apply from: "${rootDir}/gradle/publications.gradle"
 
 repositories {
+	mavenLocal()
```
remove this

gregw committed:
Merge remote-tracking branch 'webtide/main' into JettyCoreHttpHandlerAdapter

lachlan-roberts committed:
Changes from review.

Signed-off-by: Lachlan Roberts <lachlan.p.roberts@gmail.com>

lachlan-roberts committed:
Changes from review.

Signed-off-by: Lachlan Roberts <lachlan.p.roberts@gmail.com>

gregw committed:
Merge remote-tracking branch 'webtide/main' into JettyCoreHttpHandlerAdapter

gregw committed:
updated to jetty 12.0.7

gregw committed:
updated to jetty 12.0.7

gregw committed:
updated to jetty 12.0.7

gregw committed:
Merge remote-tracking branch 'webtide/main' into JettyCoreHttpHandlerAdapter

gregw committed:
updated to jetty 12.0.7

gregw committed:
updated to jetty 12.0.7

gregw committed:
fixed checkstyle

gregw committed:
fixed checkstyle

gregw committed:
Merge branch 'main' into JettyCoreHttpHandlerAdapter

gregw commented:
@jhoeller Is there anything process wise that we need to do to move this PR on?   It would be good to get some review to make sure we are on the right path.

gregw committed:
Merge branch 'spring-projects:main' into JettyCoreHttpHandlerAdapter

lachlan-roberts committed:
Implement LifeCycle for JettyWebSocketClient

Signed-off-by: Lachlan Roberts <lachlan.p.roberts@gmail.com>

lachlan-roberts committed:
Implement LifeCycle for JettyWebSocketClient

Signed-off-by: Lachlan Roberts <lachlan.p.roberts@gmail.com>

lachlan-roberts committed:
additional changes from review

Signed-off-by: Lachlan Roberts <lachlan.p.roberts@gmail.com>

lachlan-roberts committed:
additional changes from review

Signed-off-by: Lachlan Roberts <lachlan.p.roberts@gmail.com>

lachlan-roberts committed:
Merge pull request #2 from webtide/JettyCoreHttpHandlerAdapter-WebSocket

Update spring Jetty websocket usage

poutsma commented:
@gregw Hi Greg, sorry for the recent lack of feedback. I plan to take a look at this PR soon, and hopefully we can merge it in 6.2 M2 or M3.

gregw commented:
@poutsma Would be good to get at least an initial review, as I'm sure there is some rough edges that need to be knocked off.    Giving us a few comments will get us working on this and updating... plus we want to flow the work through springboot as well.

poutsma commented:
@gregw Fair point. I had to take some unforeseen medical leave, but plan to focus on this PR this coming week.

gregw commented:
> @gregw First off, thank you for submitting this PR, as well as your patience with regards to us handling it. The PR seems to be quite usable, and I am sure it will make it into 6.2.

Awesome!
 
> Before getting into any specific details, my main question is why do `JettyCoreServerHttpRequest` and `JettyCoreServerHttpResponse` implement the `ServerHttpRequest` and `ServerHttpResponse` interfaces, and not to extend the abstract base classes? It appears that a lot of changes in the PR derive from this decision (for instance `getNativeRequest` is moved from the abstract base class to the interface). While it can be argued that this refactoring is desirable, it seems unrelated and makes this PR harder to review.

Very good question.  I cannot recall and looking at the code now, I see no particular reason, at least not for the request. 
 
> How should we proceed? If you prefer, I can take the PR from here, and get back you when I have questions. Or I can provide a more thorough review.

I'm flexible.  I think it would be good to get you closely involved as I'm sure there a spring things that I'm doing wrong and working together would be a good way to communicate knowledge.   So how about this as an approach:
 
1. I'll have a go at reacting to your first feedback: i.e. try to use the abstract base classes.  That will refresh my brain on this branch (it has been a while since I looked).
2. You can then take over it for a while, changing anything you don't like or that is wrong
3. I can stay involved, helping out as you require and reviewing your changes with a Jetty eye.

gregw committed:
Use abstract request/response classes

gregw committed:
Use abstract request/response classes

lachlan-roberts committed:
Suppress NullAway warnings for Jetty WebSocket classes

Signed-off-by: Lachlan Roberts <lachlan.p.roberts@gmail.com>

lachlan-roberts committed:
Suppress NullAway warnings for Jetty WebSocket classes

Signed-off-by: Lachlan Roberts <lachlan.p.roberts@gmail.com>

lachlan-roberts committed:
fix other checkstyle warnings

Signed-off-by: Lachlan Roberts <lachlan.p.roberts@gmail.com>

lachlan-roberts committed:
fix other checkstyle warnings

Signed-off-by: Lachlan Roberts <lachlan.p.roberts@gmail.com>

poutsma commented:
> I'm flexible. I think it would be good to get you closely involved as I'm sure there a spring things that I'm doing wrong and working together would be a good way to communicate knowledge. So how about this as an approach:
> 
> 1. I'll have a go at reacting to your first feedback: i.e. try to use the abstract base classes.  That will refresh my brain on this branch (it has been a while since I looked).
> 2. You can then take over it for a while, changing anything you don't like or that is wrong
> 3. I can stay involved, helping out as you require and reviewing your changes with a Jetty eye.

Sounds good to me. It looks like you committed step 1 (d6986ee), so I will get started with step 2. Unless there are more changes forthcoming?

poutsma committed:
Polishing

gregw commented:
@poutsma my step 1 is pretty broken.   I think it need more work.  stand by....

gregw committed:
backed out abstract response change
upgrade jetty

gregw committed:
backed out abstract response change
upgrade jetty

gregw commented:
@poutsma I've backed out the abstract response changes for now, as they broke a lot of tests.  So I need to understand the difference and try again.  Probably tomorrow now...

gregw committed:
Merge remote-tracking branch 'origin/main' into JettyCoreHttpHandlerAdapter

# Conflicts:
#	framework-platform/framework-platform.gradle

gregw committed:
Steps towards abstract response class

gregw committed:
Steps towards abstract response class

lachlan-roberts committed:
add explicit dependency on jetty-websocket-api

Signed-off-by: Lachlan Roberts <lachlan.p.roberts@gmail.com>

lachlan-roberts committed:
add explicit dependency on jetty-websocket-api

Signed-off-by: Lachlan Roberts <lachlan.p.roberts@gmail.com>

poutsma commented:
> @poutsma I've backed out the abstract response changes for now, as they broke a lot of tests. So I need to understand the difference and try again. Probably tomorrow now...

@gregw No worries, just let me know when you want me to take a look.

gregw committed:
checkstyle

gregw committed:
checkstyle

gregw committed:
updated response to use the abstract

gregw committed:
updated response to use the abstract

gregw committed:
updated response to use the abstract

gregw committed:
updated response to use the abstract

gregw committed:
Merge remote-tracking branch 'refs/remotes/webtide/JettyCoreHttpHandlerAdapter' into JettyCoreHttpHandlerAdapter

poutsma committed:
Replace flatMap(...,1) with concatMap

poutsma commented:
@gregw I have made a first pass at reviewing the PR, and so far have made two changes:
* I have reverted all unnecessary changes to `ServerHttpRequest`, `ServerHttpRequest` and subtypes, and simplified `JettyCoreHttpHandlerAdapter`, `JettyCoreServerHttpRequest` and `JettyCoreServerHttpResponse`, see [here](https://github.com/poutsma/spring-framework/commit/63e772c3eaa91b15f557a200ac2ebcfa6f4a0ddc).
* I have replaced `flatMap(mapper, 1)` with `concatMap(mapper)` as these are semantically equivalent, and we use `concatMap` elsewhere, see [here](https://github.com/poutsma/spring-framework/commit/d6777c47e2e1be2765dc5d5cbd35db2757dab1f5).

Regardless of whether the changes above have been applied, the main test failure I see is `MultipartWebClientIntegrationTests`.  Next week I will take a look at why these fail. I will also take a look at the websocket changes, as I haven't done so yet.

poutsma committed:
Make copy of headers map

poutsma committed:
Introduce JettyDataBuffer

poutsma committed:
Polish JettyDataBuffer

poutsma committed:
Polishing

poutsma commented:
I have committed several other changes, see https://github.com/poutsma/spring-framework/commits/gh-32097/.

The only test that fails is the `MultipartWebClientIntegrationTests`. It does not fail in the IDE, making it a bit harder to reproduce. This is the server-side stack trace I get:

```
java.lang.IllegalArgumentException: demand pending
    at org.eclipse.jetty.server.internal.HttpChannelState$ChannelRequest.demand(HttpChannelState.java:955) ~[jetty-server-12.0.10.jar:12.0.10]
    at org.eclipse.jetty.io.content.ContentSourcePublisher$SubscriptionImpl.process(ContentSourcePublisher.java:124) ~[jetty-io-12.0.10.jar:12.0.10]
    at org.eclipse.jetty.util.thread.SerializedInvoker$Link.run(SerializedInvoker.java:191) [jetty-util-12.0.10.jar:12.0.10]
    at org.eclipse.jetty.server.internal.HttpConnection$DemandContentCallback.succeeded(HttpConnection.java:680) [jetty-server-12.0.10.jar:12.0.10]
    at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:99) [jetty-io-12.0.10.jar:12.0.10]
    at org.eclipse.jetty.io.SelectableChannelEndPoint$1.run(SelectableChannelEndPoint.java:53) [jetty-io-12.0.10.jar:12.0.10]
    at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:979) [jetty-util-12.0.10.jar:12.0.10]
    at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.doRunJob(QueuedThreadPool.java:1209) [jetty-util-12.0.10.jar:12.0.10]
    at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1164) [jetty-util-12.0.10.jar:12.0.10]
    at java.base/java.lang.Thread.run(Thread.java:840) [?:?]
```

@gregw Does that ring any bell with you, or are we doing something wrong in the Spring code?

I have also asked @rstoyanchev and @simonbasle to review (my branch of) the PR, so comments from them might be forthcoming.

The next Spring Framework milestone (6.2.0-M4) is on June 13th. I am not sure if we will have resolved the test above by that point, so this PR might not make that milestone. The milestone after that is scheduled for July 11th.

poutsma committed:
Polishing

simonbasle commented on spring-webflux/src/main/java/org/springframework/web/reactive/socket/adapter/JettyWebSocketSession.java:
```diff
@@ -51,52 +71,77 @@ public JettyWebSocketSession(Session session, HandshakeInfo info, DataBufferFact
+				boolean demand = false;
+				this.lock.lock();
+				try {
+					this.requested += n;
```
this is more problematic, as `n` can be an "unbounded request" (`Integer.MAX_VALUE`) so `this.requested` can overflow to negative.

simonbasle replied:
```suggestion
					this.requested += n;
					if (this.requested < 0L) {
						this.requested = Long.MAX_VALUE;
					}
```

gregw replied:
Using `Math.addExact(long,long)` is probably best.

lachlan-roberts replied:
I updated with the change suggested by @simonbasle not the `Math.addExact`. So once it reaches `Long.MAX_VALUE` it is considered an "unbounded request" and no longer decrements `this.requested` when demanding.

simonbasle commented on spring-webflux/src/main/java/org/springframework/web/reactive/socket/adapter/JettyWebSocketSession.java:
```diff
@@ -51,52 +71,77 @@ public JettyWebSocketSession(Session session, HandshakeInfo info, DataBufferFact
+				try {
+					this.requested += n;
+					if (!this.awaitingMessage && this.requested > 0) {
+						this.requested--;
```
I'm not entirely sure how it translates in terms of `getDelegate().demand()` calls, but if the requested amount is unbounded, this decrement shouldn't happen.

with the above suggested change, unbounded request amount (`n == Integer.MAXVALUE` at any point) leads to `this.requested == Long.MAX_VALUE` so if that's the case the `this.requested--` decrement should be skipped.

lachlan-roberts replied:
done.

simonbasle commented on spring-webflux/src/main/java/org/springframework/web/reactive/socket/adapter/JettyWebSocketSession.java:
```diff
@@ -51,52 +71,77 @@ public JettyWebSocketSession(Session session, HandshakeInfo info, DataBufferFact
+			}
+			this.awaitingMessage = false;
+			if (this.requested > 0) {
+				this.requested--;
```
same as above, this should take the "unbounded request" scenario (in the Reactive Streams sense) into consideration

lachlan-roberts replied:
done.

gregw commented:
My apologies for being slow now... I've seen your reviews and will respond in the next few days.

rstoyanchev reviewed (commented):
There are a number of additional commits in the branch referenced from https://github.com/spring-projects/spring-framework/pull/32097#issuecomment-2151849302, so do have a look there for the latest.

I've added a few comments and questions below. The only other thing to mention is we'll need a non-reactive variant of `JettyWebSocketClient` in the `spring-websocket` module for those who prefer not to use or have `spring-webflux` on the classpath.

gregw commented on spring-webflux/src/main/java/org/springframework/web/reactive/socket/adapter/JettyWebSocketSession.java:
```diff
@@ -36,13 +43,26 @@
  */
-public class JettyWebSocketSession extends AbstractListenerWebSocketSession<Session> {
+@SuppressWarnings("NullAway")
+public class JettyWebSocketSession extends AbstractWebSocketSession<Session> {
```
@lachlan-roberts 
can you look at the websocket part of these review

lachlan-roberts replied:
The JettyWebSocket API uses a demand model instead of suspend/resume.

The `AbstractListenerWebSocketSession` requires us to implement `suspendReceiving()` and `resumeReceiving()` so we're better off not extending that class.

gregw replied:
Ah yes! Suspend/resume is intrinsically racy and best avoided.

gregw commented on spring-web/src/testFixtures/java/org/springframework/web/testfixture/http/server/reactive/bootstrap/JettyCoreHttpServer.java:
```diff
@@ -0,0 +1,100 @@
+		this.jettyServer.setHandler(createHandlerAdapter());
+
+		// TODO: We don't actually want the upgrade handler but this will create the WebSocketContainer.
+		//  This requires a change in Jetty.
```
@lachlan-roberts  made the change in jetty

lachlan-roberts replied:
Change has already been made in Jetty so I updated with the proper method to use and removed the TODO.
see https://github.com/jetty/jetty.project/issues/11356

gregw commented on spring-web/src/main/java/org/springframework/http/server/reactive/JettyCoreServerHttpResponse.java:
```diff
@@ -0,0 +1,315 @@
+
+	@Override
+	protected Mono<Void> writeAndFlushWithInternal(Publisher<? extends Publisher<? extends DataBuffer>> body) {
+		return Flux.from(body).flatMap(this::writeWithInternal, 1).then();
```
Jetty writes at this level are non-buffering.  Any aggregation that might require flushing is done at a higher level.     Now, we might actually buffer (e.g. if sitting in a flow control congested HTTP/2 stream), but if so, then flushing is not going to help.

The only time we really need a flush semantic is if there is no content at all, then we will write an empty buffer, so that at least the HTTP response headers are generated etc. So perhaps this method needs to detect if the body was actually empty, in which case an empty buffer should be written

gregw commented on spring-web/src/main/java/org/springframework/http/server/reactive/JettyCoreServerHttpResponse.java:
```diff
@@ -0,0 +1,315 @@
+
+		private final long length;
+
+		private long totalRead = 0;
```
Good idea.  It might be in a slightly different form... probably something like:
```java
			Content.copy(Content.Source.from(channel, position, count), this.response, callback);
```
I'll try to get that in this months release.

gregw commented:
@poutsma Our branches have diverged as I made the changes to use `AbstractServerHttpRequest` and `AbstractServerHttpResponse` in my branch and you forked before that.     Let me see if I can rebase yours onto mine....

gregw committed:
Merge remote-tracking branch 'poutsma/gh-32097' into JettyCoreHttpHandlerAdapter

# Conflicts:
#	spring-web/src/main/java/org/springframework/http/server/reactive/DefaultServerHttpRequestBuilder.java
#	spring-web/src/main/java/org/springframework/http/server/reactive/JettyCoreHttpHandlerAdapter.java
#	spring-web/src/main/java/org/springframework/http/server/reactive/JettyCoreServerHttpRequest.java
#	spring-web/src/main/java/org/springframework/http/server/reactive/JettyCoreServerHttpResponse.java
#	spring-webflux/src/main/java/org/springframework/web/reactive/socket/adapter/JettyWebSocketHandlerAdapter.java
#	spring-webflux/src/main/java/org/springframework/web/reactive/socket/adapter/JettyWebSocketSession.java
#	spring-websocket/src/main/java/org/springframework/web/socket/adapter/jetty/JettyWebSocketHandlerAdapter.java

gregw committed:
Merge remote-tracking branch 'poutsma/gh-32097' into JettyCoreHttpHandlerAdapter

# Conflicts:
#	spring-web/src/main/java/org/springframework/http/server/reactive/DefaultServerHttpRequestBuilder.java
#	spring-web/src/main/java/org/springframework/http/server/reactive/JettyCoreHttpHandlerAdapter.java
#	spring-web/src/main/java/org/springframework/http/server/reactive/JettyCoreServerHttpRequest.java
#	spring-web/src/main/java/org/springframework/http/server/reactive/JettyCoreServerHttpResponse.java
#	spring-webflux/src/main/java/org/springframework/web/reactive/socket/adapter/JettyWebSocketHandlerAdapter.java
#	spring-webflux/src/main/java/org/springframework/web/reactive/socket/adapter/JettyWebSocketSession.java
#	spring-websocket/src/main/java/org/springframework/web/socket/adapter/jetty/JettyWebSocketHandlerAdapter.java

gregw committed:
Merge remote-tracking branch 'origin/main' into JettyCoreHttpHandlerAdapter

gregw commented:
Actually, I'm not exactly sure why our branches were seen as divergent.   I've merged yours back to mine and made sure that there are no differences, and also merged to lastest origin/main.
I'm not 100% sure the merge is correct, so we may want to proceed with your branch, but at least we can see your changes in this PR and I can work with it as well.   Now looking at the  test failure....

gregw committed:
addExact

gregw commented:
@poutsma 
> The only test that fails is the `MultipartWebClientIntegrationTests`. It does not fail in the IDE, making it a bit harder to reproduce. This is the server-side stack trace I get:
> 
> ```
> java.lang.IllegalArgumentException: demand pending
>   ...
> ```
> 
> @gregw Does that ring any bell with you, or are we doing something wrong in the Spring code?

I think it is likely a problem with the jetty ContentSourcePublisher, for which we already have a PR in review: https://github.com/jetty/jetty.project/pull/11849

> The next Spring Framework milestone (6.2.0-M4) is on June 13th. I am not sure if we will have resolved the test above by that point, so this PR might not make that milestone. The milestone after that is scheduled for July 11th.

Agreed we are not making the 13th.   We have a jetty release at the end of the month, so I will get #11849 merged for that and also look at change @rstoyanchev made.

@lachlan-roberts can you look at the websocket reviews before end of month, so that any changes needed in next jetty release can be included.  E.g. anything needed to get the Container without the Handler

gregw committed:
doCommit not required

gregw commented on spring-web/src/main/java/org/springframework/http/server/reactive/JettyCoreServerHttpResponse.java:
```diff
@@ -0,0 +1,315 @@
+			}
+		}.iterate();
+
+		return doCommit(() -> Mono.fromFuture(callback));
```
Not necessary. removed

lachlan-roberts committed:
PR #32097 - changes to address reviews for websocket

Signed-off-by: Lachlan Roberts <lachlan.p.roberts@gmail.com>

gregw commented:
The `MultipartWebClientIntegrationTests` test did not fail running from command line with jetty 12.0.11-SNAPSHOT

poutsma commented:
> The `MultipartWebClientIntegrationTests` test did not fail running from command line with jetty 12.0.11-SNAPSHOT

Great! If I understand your other comment correctly, Jetty 12.0.11 will be out by the end of June, which should give us enough time to get this PR into 6.2.0-M5.

gregw committed:
doCommit not required

lachlan-roberts committed:
fix checkstyle errors

Signed-off-by: Lachlan Roberts <lachlan.p.roberts@gmail.com>

lachlan-roberts committed:
fix failure for CookieIntegrationTests.partitionedAttributeTest

Signed-off-by: Lachlan Roberts <lachlan.p.roberts@gmail.com>

lachlan-roberts commented:
@poutsma @rstoyanchev the build of this branch is now passing all tests with the staged release of jetty-12.0.11, which should be released very soon.

simonbasle commented:
@lachlan-roberts @gregw I see the artifacts appear to be on [Maven Central as of June 28](https://central.sonatype.com/artifact/org.eclipse.jetty/jetty-bom/12.0.11/versions), but I see no announcements and the release-tracking issue [`jetty/jetty.project/issues/11980`](https://www.github.com/jetty/jetty.project/issues/11980) is not closed. Can you confirm if we should be waiting for anything before merging?

I've retrieved Arjen's branch locally, made a couple polishes (fixing duplicate start/stop methods in `JettyWebSocketClient` notably), rebased on the latest main and all the tests ran successfully 👍

gregw commented:
@simonbasle The release is out in maven central and there is no taking it back.  We have some people on vacation, hence the issue process has not yet completed.

simonbasle commented:
@lachlan-roberts unless I've missed something, you haven't yet signed the CLA have you?

lachlan-roberts commented:
@simonbasle yes I have already signed the CLA.

JohnNiang commented on spring-web/src/main/java/org/springframework/http/server/reactive/DefaultServerHttpRequestBuilder.java:
```diff
@@ -70,7 +71,8 @@ public DefaultServerHttpRequestBuilder(ServerHttpRequest original) {
 		this.uri = original.getURI();
-		this.headers = new HttpHeaders(original.getHeaders());
+		// original headers can be immutable, so create a copy
+		this.headers = new HttpHeaders(new LinkedMultiValueMap<>(original.getHeaders()));
```
If the headers is copied into LinkedMultiValueMap directly, it will break the case insensitive behavior of header names.

e.g.: If the headers contain an entry `content-type: application/json`(HTTP2 requires header names with lowercase), and the result of `headers.getContentType()` or `headers.getFirst("Content-Type")` will be null forever.

bclozel replied:
Can you raise this as a new issue?

simonbasle replied:
indeed, thanks for the feedback! I've created issue #33666 to track and fix this in RC2
